Components/main_system.cpp

```diff
@@ -0,0 +1,47 @@
+/**
+ ******************************************************************************
+ * File Name          : main_system.cpp
+ * Description        : This file acts as an interface supporting CubeIDE Codegen
+    while having a clean interface for development.
+ ******************************************************************************
+*/
+/* Includes -----------------------------------------------------------------*/
+#include "SystemDefines.hpp"
+#include "UARTDriver.hpp"
+
+// Tasks
+#include "CubeTask.hpp"
+
+/* Drivers ------------------------------------------------------------------*/
+namespace Driver {
+    UARTDriver uart2(USART2);
+}
+
+/* Interface Functions ------------------------------------------------------------*/
+/**
+ * @brief Main function interface, called inside main.cpp before os initialization takes place.
+*/
+void run_main() {
+    // Init Tasks
+    CubeTask::Inst().InitTask();
+
+    // Print System Boot Info : Warning, don't queue more than 10 prints before scheduler starts
+    CUBE_PRINT("\n-- CUBE SYSTEM --\n");
+    CUBE_PRINT("System Reset Reason: [TODO]\n"); //TODO: System reset reason can be implemented via. Flash storage
+    CUBE_PRINT("Current System Heap Use: %d Bytes\n", xPortGetFreeHeapSize());
+    CUBE_PRINT("Lowest Ever Heap Size: %d Bytes\n\n", xPortGetMinimumEverFreeHeapSize());
+
+    // Start the Scheduler
+    // Guidelines:
+    // - Be CAREFUL with race conditions after osKernelStart
+    // - All uses of new and delete should be closely monitored after this point
+    osKernelStart();
+
+    // Should never reach here
+    CUBE_ASSERT(false, "osKernelStart() failed");
+
+    while (1)
+    {
+        osDelay(100);
```

Comment:
It's mainly so if it ever get's here (somehow, I think it'd have to be EM radiation corrupting something), other tasks could 'run' instead of the loop taking full control, though unlikely in either case. 

If we need the instruction space we can remove it tho.

Reply:
Good call, I used HAL_NVIC_SystemReset() so it's more direct (since if the assert failed above too it's pretty bad). I'm half and half on keeping the osDelay there, but I think we should keep it just to avoid spamming the NVIC if it ever happens.